### PR TITLE
[RDY] Make the entirety of the command args configurable

### PR DESCRIPTION
**Pull request by robbyoconnor**
Allow for more flexibility in how certbot is executed.

### CODE REVIEW

**Comment by gronke** [comment]
LGTM 👍

File: defaults/main.yml
Comment (gronke) [inline defaults/main.yml]: I think having all flags in a list is an improvement to the current solution. Anyway there are a few parameters that are mandatory for this role to work. Maybe having a variable `letsencrypt_certbot_default_args` that is combined with the user provided ones is easier to use than having to remember all flags when the intention was only to add one.

``` yaml
letsencrypt_default_certbot_args:
  - "certonly"
  - "--text""
  ...
```

and within the role

``` yaml
- set_fact: _letsencrypt_combined_args="{{ letsencrypt_default_certbot_args + letsencrypt_certbot_args }}"

- set_fact: _letsencrypt_certbot_args="{{ _letsencrypt_combined_certbot_args + ['--renew-by-default'] }}"
  when: letsencrypt_force_renew == true

- set_fact: _letsencrypt_certbot_args="{{ _letsencrypt_combined_certbot_args + ['--keep-until-expiring'] }}"
  when: letsencrypt_force_renew != true
```

File: defaults/main.yml
Comment (gronke) [inline defaults/main.yml]: The list style used here is indeed a bit better to read than the one currently used in this role, but I would prefer not to mix the styles in one role.

#### before

``` yaml
letsencrypt_certbot_args: ["certonly",
  "--text",
 ...
```

#### after

``` yaml
letsencrypt_certbot_args:
  - "certonly"
  - "--text"
 ...
```

Comment (jaywink) [reply defaults/main.yml]: I kind of prefer the `["foo", "bar"]` syntax myself over the more common (?) way of doing lists in YAML, I think for cli params it is a bit clearer, less confusing looking `-`'s around :)

File: defaults/main.yml
Comment (jaywink) [reply defaults/main.yml]: I do agree having everything neatly in a list is an improvement. However it does confuse the way the role works since some of the things right now are declared via variables, which dynamically affect the args contents, but at the same time the whole list of args can just be declared at once. I do find these two ways of configuring do create some confusion on how the role should be used.

@robbyoconnor may I ask what is the use case of allowing to override the default arguments currently provided? As even now you can provide extra arguments, so it must be the need to change some of the current hard coded defaults? Could these be provided as separate configuration flags like some of the previous ones?

The reason I like the more easier to understand configuration flags is that the role then hides the complexity of certbot, allowing the user not to need to understand what flags need to be passed in. If we're missing some use case here that needs extra flags passed in to replace the current default, I would personally prefer overriding them via some configuration item.

Your thoughts on this @gronke ?

Comment (gronke) [reply defaults/main.yml]: As long as the arguments are order independent I don't see a problem having a second variable that the casual user of this role does not touch. As soon as someone attempts to override these defaults the configuration potentially becomes incompatible with future role or certbot updates, which is why I would not recommend to override that variable from outside other than for testing purposes.

It's neat to have every flag in a different line, so that changes become easier to track in git.

File: defaults/main.yml
Comment (gronke) [reply defaults/main.yml]: I don't mind to change that everywhere in this role. Just mixing different styles is confusing.

Personally I prefer YAML lists because
- the line defining the variable name is not affected when the first argument changes
- entries are easier to read when all are beginning at the same indention level
- adding an entry to the end does not affect the previously last line (removal of `]` at the end of the last line)

**Comment by robbyoconnor** [comment]
This satisfied an itch I had -- namely wanting control over how it's run...just contributing upstream because FOSS :)

File: defaults/main.yml
Comment (robbyoconnor) [reply defaults/main.yml]: The default is exactly as you had it @jaywink -- so if people don't care, they don't have to change it. This was an itch I had...

File: defaults/main.yml
Comment (robbyoconnor) [reply defaults/main.yml]: What's the verdict here?

**Comment by gronke** [comment]
@robbyoconnor thank you for sharing! Would you to rename the variable containing the default flags like [mentioned in this comment](15#discussion_r79638022).

File: defaults/main.yml
Comment (gronke) [reply defaults/main.yml]: @robbyoconnor it's just perfect as it is. Later we can figure out which list style is the better one - YAML leaves too much room for opinions 😼

**Comment by robbyoconnor** [comment]
@gronke, @jaywink: sure will do. Would you mind if I also enabled Diffie Hellman parameters? This helps to make the Qualsys higher which means more security.

File: defaults/main.yml
Comment (robbyoconnor) [reply defaults/main.yml]: I like it this way personally.

Comment (robbyoconnor) [reply defaults/main.yml]: Will update the default params to another one... my issue is that having them together is better...for me at least -- I copy/paste the default params and change what I need to and delete what I don't -- so seeing what's actually set and changing things is easiest for me...

One way we can make this easier: 
- an option to run in standalone mode
- an option to generate the apache config 
- an option to generate diffie hellman parameters with 4096 bits (security is important to me)

**Comment by robbyoconnor** [comment]
Changes have been pushed @gronke, @jaywink

**Comment by robbyoconnor** [comment]
README updated as well.

**Comment by gronke** [comment]
> Would you mind if I also enabled Diffie Hellman parameters?

I prefer using this role exclusively to obtain certificates from LetsEncrypt. Generating Diffie-Hellman parameters is also important, but a step independent of LetsEncrypt. 

Depending on your operating system, hardware and random-number-generator generating large dhparams can take quite a while and Ansible is not very good in giving feedback about running shell tasks. Putting this job into the LetsEncrypt role could easily confuse people.

When you for example decide to switch from this role to another that generates self-signed certificates you probably still want the dhparams to be generated. So for me that is a task that is not related to letsencrypt. I'm using

``` yaml
pre_tasks:
  - name: Diffie-Hellman parameter is generated
    shell: /usr/bin/openssl dhparam -out /etc/ssl/certs/dhparam.pem 4096
    creates: /etc/ssl/certs/dhparam.pem
roles:
  - role: jaywink.letsencrypt
    letsencrypt_domain: example.com
    letsencrypt_email: admin@example.com
```

> Will update the default params to another one... my issue is that having them together is better...for me at least -- I copy/paste the default params and change what I need to and delete what I don't -- so seeing what's actually set and changing things is easiest for me...

Mostly this role abstracts the parameters passed to Certbot behind feature toggles. The problem with having one list containing the default options as well custom extra options is that you need to pass them all if you just want to add a single one.

The reason behind the extra parameters was to be able to pass additional flags to Certbot, like for example `--test-cert`, not to tweak the default ones.

As mentioned, this role is built to abstract flags behind configuration toggles. In the case you described it would be much easier to just run certbot via shell/command task than using this role. An `install_only` option would be nice for you, so that you can make use of this role for installing the client.

File: defaults/main.yml
Comment (gronke) [reply defaults/main.yml]: > I like it this way personally.

Cool. After hearing that @jaywink is also in favor of this notation, I'm happy to open a PR to change it in other places too 😄

**Comment by robbyoconnor** [comment]
I mean this still does this -- you don't HAVE TO override anything -- look at the defaults I set.. I want to not have to ssh in to run things if i can help it. 

Having this flexibility is nice (I know I like it and I'm sure there's probably another person somewhere who would as well)

File: defaults/main.yml
Comment (robbyoconnor) [inline defaults/main.yml]: @gronke these are the defaults as you had them -- they don't have to be touched and will work as you originally intended.

**Comment by robbyoconnor** [comment]
Don't merge this (if you were going to) need to fix it. I seem to have messed up the facts. If changes aren't accepted here, I may fork this(not a threat at all, I hope it does not come off as such -- just seems we have competing visions)... I also wanna add an automatic cronjob to renew.

File: defaults/main.yml
Comment (gronke) [reply defaults/main.yml]: Maybe the name `letsencrypt_certbot_args` is a bit misleading. It should give the ability to pass additional flags to certbot. In the given case you'd need to also pass `["--standalone", "--expand", "--my-extra-flag"]` instead of just `["--my-extra-flag"]`. This could break upwards compatibility of roles in use passing extra flags, when maintainers did not notice the change or in consequence not allow us to make any further changes. It becomes extra hard when we want to keep support for different versions of Certbot.

We could rename the `letsencrypt_certbot_args` to `letsencrypt_certbot_extra_args` and add two more variables to the configuration

``` yaml
letsencrypt_certbot_expand: true
letsencrypt_certbot_mode: "standalone"
# certbot custom arguments -- see: https://certbot.eff.org/docs/using.html
letsencrypt_certbot_extra_args: []
```

and two more tasks to `cert.yml`

``` yaml
- set_fact: _letsencrypt_certbot_args="{{ _letsencrypt_certbot_args + ['--standalone'] }}"
  when: letsencrypt_certbot_mode == "standalone"
- set_fact: _letsencrypt_certbot_args="{{ _letsencrypt_certbot_args + ['--expand'] }}"
  when: letsencrypt_certbot_expand == true
```

**Comment by gronke** [comment]
:+1: for the automated crobjob. I'd love to see a pull-request with this. Good catch!

@robbyoconnor sorry that I'm so picky with your changes 😼  Our visions are not that different so far - your points are all valid. We just need to figure out how we can keep the role maintainable for the future and make it easy and intuitive to use.

**Comment by jaywink** [comment]
Phew, a lot of discussion :) Sorry for not involving myself that much, was on a family trip almost for a week so have had little time to think things through properly. Will go through the comments asap.

**Comment by jaywink** [comment]
Firstly, thanks @robbyoconnor for your effort in providing this change and @gronke for tireless review :)

I'm fine with keeping `letsencrypt_certbot_args` as that is what the "extra args" variable is already called. This would mean no backwards compatibility changes for anyone who is already using this role. If you change it, we just need to make sure to mention it clearly in the changelog. No strong opinion on this.

Otherwise I agree completely to what @gronke has been saying. The default args should be in their own variable (as they are now in the PR) and another variable to allow passing "extra" easily without needing to change the default. I would go as far as not mentioning the defaults variable in the readme "required" section. It can be mentioned somewhere later as a "if you like, you can..." kind of thing.

We should provide "feature flags" to change functionality instead of relying everybody to have to know certbot usage (which imho isn't that well explained for your basic sysadmin to understand). We should also try to attack the most common cases with the defaults.

This PR will provide role users the possibility to just set everything, which is a nice power user addition.

> ...and add two more variables to the configuration

I'd rather handle these as other PR's. Dropping standalone with a boolean would need some other changes too afaik? Like specifying `--web-root` as per #5 .

> I also wanna add an automatic cronjob to renew. 

Also would be happy to accept an cronjob flag too, maybe default off. 👍

File: README.md
Comment (jaywink) [inline README.md]: As mentioned in the larger comment, this shouldn't be a "required" since it's not really - the defaults _are_ there.

File: defaults/main.yml
Comment (jaywink) [inline defaults/main.yml]: If you keep this YAML list style, you should be able to drop the `"`'s afaict, though I guess it doesn't matter that much.

File: defaults/main.yml
Comment (jaywink) [reply defaults/main.yml]: The `--standalone` and `--expand` should be in the `letsencrypt_certbot_default_args` variable imho.

File: tasks/cert.yaml
Comment (jaywink) [inline tasks/cert.yaml]: This should use `_letsencrypt_combined_certbot_args`.

**Review by jaywink — CHANGES_REQUESTED** [review]

File: tasks/cert.yaml
Comment (robbyoconnor) [reply tasks/cert.yaml]: Okay.

File: defaults/main.yml
Comment (robbyoconnor) [reply defaults/main.yml]: I'll drop the double quotes -- do you have a strong preference?

File: README.md
Comment (robbyoconnor) [reply README.md]: Okay -- I'll fix this!

**Comment by robbyoconnor** [comment]
@jaywink what about a flag to generate diffie-hellman params? I do think I agree with @gronke -- they do take ~10-15 mins to run and would stall the playbook...

File: defaults/main.yml
Comment (jaywink) [reply defaults/main.yml]: No strong preference, it's just that they're not required.

```
items:
  - foo 

items:
  - "foo"

items: ["foo"]
```

As said, for a list of cli params, I think the last one is clearer. But for other YAML lists, I prefer the first one. It's probably better stay consistent to one or the other in this role. Since the first one is the more common one in Ansible roles afaik, I'd say lets go with that one, without the quotes if possible.

**Comment by jaywink** [comment]
> what about a flag to generate diffie-hellman params? I do think I agree with @gronke -- they do take ~10-15 mins to run and would stall the playbook...

I don't have a problem with taking this role a little more in the direction of "letsencrypt with strong SSL setup" rather than just "letsencrypt". So sure, if you provide the by default off dh params task and the glue for both Apache and nginx to use them - I'd happily include that in the role, unless @gronke objects ;) As long as the defaults serve the "majority simple cases", with configuration items for additional setup.

Once this current PR is merged I'll do a release 💥

**Comment by robbyoconnor** [comment]
I'll do this either Sat or Sat...I wanna sleep right now.

**Comment by gronke** [comment]
haha.. I don't see a direct link between LetsEncrypt and Diffie-Hellman parameters. You could want to use your certificate with different dhparams on different service, which directly makes things a bit more complicated.

@jaywink @robbyoconnor This role is dedicated to you: [gronke.dhparam](https://github.com/gronke/ansible-dhparam) 😼

**Comment by robbyoconnor** [comment]
@gronke you're amazing

File: defaults/main.yml
Comment (gronke) [reply defaults/main.yml]: Yes, we can move that out here when we introduce a toggle for the standalone mode in a later pull-request.

**Comment by gronke** [comment]
Reviewing our comments, there are two more changes necessary before we can merge:
- Move `--standalone` and `--expand` into letsencrypt_certbot_default_args
- Update Readme (letsencrypt_certbot_args is not required)

@robbyoconnor would you mind if I create some PRs from the ideas we had during the review?

**Comment by robbyoconnor** [comment]
@gronke against my fork or here...No I don't mind at all in either case. I made some PRs against your nginx playbook -- your playbooks are amazing :heart:

File: defaults/main.yml
Comment (robbyoconnor) [reply defaults/main.yml]: I'm just now getting to this :smile:

Comment (robbyoconnor) [reply defaults/main.yml]: @gronke  you might wanna hold off on your PR until mine is merged -- I can probably handle your PR in mine.

**Comment by gronke** [comment]
@robbyoconnor The role now officially has more Pull-Requests than Downloads 🎉 I'm happy to see you have fun extending it!

I'll hold back my PRs and rebase against your changes once they are mered. (Btw. I'm working on running certbot as unprivileged user. Probably that can only be elegantly solved with a custom Ansible module because certbot-auto itself [insists on being root](https://github.com/certbot/certbot/blob/master/certbot-auto#L84))

**Comment by robbyoconnor** [comment]
@gronke certbot needs to be root....

**Comment by robbyoconnor** [comment]
I'm going to now edit the README :)

**Comment by robbyoconnor** [comment]
Mostly done -- just testing things -- I'll squash when I'm done.

**Comment by robbyoconnor** [comment]
@gronke, I couldn't get it to work...but this is enough to be flexible enough for me...I can modify what's passed and I'm happy.

**Comment by robbyoconnor** [comment]
@jaywink as well I guess

File: defaults/main.yml
Comment (gronke) [inline defaults/main.yml]: This variable is defined via `_set_fact` task later, so it can be omitted from the defaults file.

File: defaults/main.yml
Comment (gronke) [inline defaults/main.yml]: It was @jaywink 's with to remain calling this variable `letsencrypt_certbot_args` to keep the role compatible with current playbook setups.

File: tasks/cert.yaml
Comment (gronke) [inline tasks/cert.yaml]: According to my previous comment, please rename `letsencrypt_certbot_extra_args` to `letsencrypt_certbot_args`

File: tasks/cert.yaml
Comment (gronke) [inline tasks/cert.yaml]: Order could be changed for better readability. e.g.

``` yaml
set_fact: _signing_skipped='{{ (certbot_skip_renewal_message in _certbot_command.stdout) and not letsencrypt_force_renew }}'
```

File: tasks/cert.yaml
Comment (gronke) [inline tasks/cert.yaml:27]: could be simplified to

``` yaml
when: letsencrypt_certbot_verbose or not (_signing_successful and _signing_skipped)
```

**Review by gronke — CHANGES_REQUESTED** [review]
I'm sorry for the confusion with the `letsencrypt_certbot_extra_args`. Left some comments with suggestions for simplification.

File: tasks/cert.yaml
Comment (robbyoconnor) [reply tasks/cert.yaml]: Didn't change these -- just switched the from == false to treating it like a boolean

File: tasks/cert.yaml
Comment (robbyoconnor) [reply tasks/cert.yaml:27]: Same deal.

File: defaults/main.yml
Comment (robbyoconnor) [inline defaults/main.yml]: I was having issues running this without defining it..

Comment (robbyoconnor) [inline defaults/main.yml]: ignore last comment

File: tasks/cert.yaml
Comment (robbyoconnor) [reply tasks/cert.yaml]: Okay will do.

File: defaults/main.yml
Comment (robbyoconnor) [reply defaults/main.yml]: Will do.

**Comment by robbyoconnor** [comment]
GIven tha radical change in #18 -- is this worth my time?

**Comment by robbyoconnor** [comment]
@gronke @jaywink -- changes made.

**Comment by robbyoconnor** [comment]
Is this gonna be merged before the release? @jaywink

**Comment by jaywink** [comment]
Sorry, since @gronke already merged your commit with his PR I though it could come through there in the next release. So soon! :)

**Comment by robbyoconnor** [comment]
Shall I close this?

**Comment by gronke** [comment]
If were sure to merge this shortly after the release with the other branch, I suggest to close this PR and snapshot the current master as last certbot release.

**Comment by robbyoconnor** [comment]
I'll let @jaywink close this if he so chooses

File: defaults/main.yml
Comment (jaywink) [inline defaults/main.yml]: This shouldn't be here afaict. `_letsencrypt_certbot_combined_args` is populated automatically later in the tasks.

File: README.md
Comment (jaywink) [inline README.md]: I'd add here to the end of the line `Please see "defaults/main.yml" what the default arguments are.`. Also, you could add `To override all the arguments to Certbot, for example to use another plugin, set them using this variable.`. This makes the comment on line 67 redundant.

File: README.md
Comment (jaywink) [inline README.md]: I'd remove this as per above comment.

**Review by jaywink — CHANGES_REQUESTED** [review]
Let's merge this in separately after all. We decided with @gronke to not merge the simp_le PR in, at least not at this moment, due to simp_le project not looking active at the moment.

So a few last changes if possible and then will merge! :)

**Comment by robbyoconnor** [comment]
Changes have been made @jaywink

File: defaults/main.yml
Comment (robbyoconnor) [reply defaults/main.yml]: Done.

File: README.md
Comment (robbyoconnor) [reply README.md]: Done.

File: README.md
Comment (robbyoconnor) [reply README.md]: :x:'d

**Review by jaywink — APPROVED** [review]

**Comment by jaywink** [comment]
Awesome, thanks! Finally merging this 💃

**Comment by gronke** [comment]
👍 🎉

**Comment by robbyoconnor** [comment]
@jaywink is a release going to be cut soon?

**Comment by jaywink** [comment]
@robbyoconnor done! :)

https://github.com/jaywink/ansible-letsencrypt/releases/tag/v0.6.0

**Pull request merged**
